diff --git a/orocos_kdl_vendor/CMakeLists.txt b/orocos_kdl_vendor/CMakeLists.txt
index 655e562..58181a6 100644
--- a/orocos_kdl_vendor/CMakeLists.txt
+++ b/orocos_kdl_vendor/CMakeLists.txt
@@ -49,6 +49,15 @@ macro(build_orocos_kdl)
   list(APPEND extra_cmake_args "-DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}")
   list(APPEND extra_cmake_args "-DENABLE_TESTS=OFF")
   list(APPEND extra_cmake_args "-DENABLE_EXAMPLES=OFF")
+  if(DEFINED EIGEN3_INCLUDE_DIR)
+    list(APPEND extra_cmake_args "-DEIGEN3_INCLUDE_DIR=${EIGEN3_INCLUDE_DIR}")
+  endif()
+  if(DEFINED Eigen3_DIR)
+    list(APPEND extra_cmake_args "-DEigen3_DIR=${Eigen3_DIR}")
+  endif()
+  if(DEFINED CMAKE_PREFIX_PATH)
+    list(APPEND extra_cmake_args "-DCMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}")
+  endif()
 
   include(ExternalProject)
 
