diff --git a/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp b/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp
index a7262bb..ab3e34f 100644
--- a/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp
+++ b/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp
@@ -234,19 +234,31 @@ rmw_ret_t PublisherData::publish(
   }
   rcutils_allocator_t * allocator = &rmw_node_->context->options.allocator;
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   // Optional shared memory buffer
   std::optional<zenoh::ZShmMut> shm_buf = std::nullopt;
+#endif
   // Optional buffer reused for serialization from the buffer pool
   std::optional<BufferPool::Buffer> pool_buf = std::nullopt;
 
   auto always_free_msg_bytes = rcpputils::make_scope_exit(
-    [&msg_bytes, allocator, &shm_buf, &pool_buf]() {
-      if (!msg_bytes || shm_buf.has_value() || (pool_buf.has_value() && pool_buf.value().data)) {
+    [&msg_bytes, allocator,
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
+     &shm_buf,
+#endif
+     &pool_buf]() {
+      if (!msg_bytes ||
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
+        shm_buf.has_value() ||
+#endif
+        (pool_buf.has_value() && pool_buf.value().data))
+      {
         return;
       }
       allocator->deallocate(msg_bytes, allocator->state);
     });
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   // Get memory from SHM buffer if available.
   if (shm && max_data_length >= shm->msgsize_threshold) {
     if (auto shm_provider = shm->get_shm_provider(*sess_)) {
@@ -269,8 +281,11 @@ rmw_ret_t PublisherData::publish(
         "rmw_zenoh_cpp", "SHM provider is not yet available, fallback to non-SHM");
     }
   }
+#endif
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   if (!shm_buf.has_value()) {
+#endif
     // Try to get memory from the serialization buffer pool.
     pool_buf = context_impl->serialization_buffer_pool()->allocate(max_data_length);
     if (pool_buf.has_value() && pool_buf.value().data) {
@@ -281,7 +296,9 @@ rmw_ret_t PublisherData::publish(
         data, "failed to allocate serialization buffer", return RMW_RET_BAD_ALLOC);
       msg_bytes = static_cast<uint8_t *>(data);
     }
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   }
+#endif
 
   RMW_CHECK_FOR_NULL_WITH_MSG(
     msg_bytes, "bytes for message is null", return RMW_RET_BAD_ALLOC);
@@ -312,9 +329,12 @@ rmw_ret_t PublisherData::publish(
     sequence_number_++, source_timestamp, entity_->copy_gid()).serialize_to_zbytes();
 
   zenoh::Bytes payload;
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   if (shm_buf.has_value()) {
     payload = zenoh::Bytes(std::move(*shm_buf));
-  } else if (pool_buf.has_value() && pool_buf.value().data) {
+  } else
+#endif
+  if (pool_buf.has_value() && pool_buf.value().data) {
     auto deleter = [buffer_pool = context_impl->serialization_buffer_pool(),
         buffer = pool_buf](uint8_t *) {
         buffer_pool->deallocate(buffer.value());
@@ -359,8 +379,10 @@ rmw_ret_t PublisherData::publish_serialized_message(
     return RMW_RET_ERROR;
   }
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   // Optional shared memory buffer
   std::optional<zenoh::ZShmMut> shm_buf = std::nullopt;
+#endif
 
   std::lock_guard<std::mutex> lock(mutex_);
 
@@ -374,6 +396,7 @@ rmw_ret_t PublisherData::publish_serialized_message(
   opts.put_options.attachment = rmw_zenoh_cpp::AttachmentData(
     sequence_number_++, source_timestamp, entity_->copy_gid()).serialize_to_zbytes();
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   // Get memory from SHM buffer if available.
   if (shm && data_length >= shm->msgsize_threshold) {
     if (auto shm_provider = shm->get_shm_provider(*sess_)) {
@@ -395,7 +418,9 @@ rmw_ret_t PublisherData::publish_serialized_message(
         "rmw_zenoh_cpp", "SHM provider is not yet available, fallback to non-SHM");
     }
   }
+#endif
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   if (shm_buf.has_value()) {
     auto msg_bytes = reinterpret_cast<char *>(shm_buf.value().data());
     memcpy(msg_bytes, serialized_message->buffer, data_length);
@@ -406,7 +431,9 @@ rmw_ret_t PublisherData::publish_serialized_message(
       source_timestamp);
 
     pub_.put(std::move(payload), std::move(opts), &result);
-  } else {
+  } else
+#endif
+  {
     std::vector<uint8_t> raw_image(
       serialized_message->buffer,
       serialized_message->buffer + data_length);
diff --git a/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp b/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp
index c54b8a8..e5ec547 100644
--- a/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp
+++ b/rmw_zenoh_cpp/src/detail/zenoh_utils.cpp
@@ -133,6 +133,7 @@ ShmContext::ShmContext(size_t msgsize_threshold)
 : msgsize_threshold(msgsize_threshold)
 {}
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
 std::optional<zenoh::SharedShmProvider> ShmContext::get_shm_provider(zenoh::Session & session)
 {
   auto maybe_provider = session.obtain_shm_provider();
@@ -141,6 +142,7 @@ std::optional<zenoh::SharedShmProvider> ShmContext::get_shm_provider(zenoh::Sess
   }
   return std::nullopt;
 }
+#endif
 
 
 ///=============================================================================
diff --git a/rmw_zenoh_cpp/src/detail/zenoh_utils.hpp b/rmw_zenoh_cpp/src/detail/zenoh_utils.hpp
index 216421f..214d98b 100644
--- a/rmw_zenoh_cpp/src/detail/zenoh_utils.hpp
+++ b/rmw_zenoh_cpp/src/detail/zenoh_utils.hpp
@@ -108,7 +108,9 @@ struct ShmContext
 
   explicit ShmContext(size_t msgsize_threshold);
 
+#if defined(Z_FEATURE_SHARED_MEMORY) && defined(Z_FEATURE_UNSTABLE_API)
   std::optional<zenoh::SharedShmProvider> get_shm_provider(zenoh::Session & session);
+#endif
 };
 
 ///=============================================================================
diff --git a/zenoh_cpp_vendor/CMakeLists.txt b/zenoh_cpp_vendor/CMakeLists.txt
index 07c8f50..d361475 100644
--- a/zenoh_cpp_vendor/CMakeLists.txt
+++ b/zenoh_cpp_vendor/CMakeLists.txt
@@ -23,7 +23,15 @@ else()
   # Note: the list of features are separated with whitespaces.
   # However, if you want to add other cargo build flags, use "$<SEMICOLON>" as separator and not ";" as the
   # latter is a list separater in cmake and hence the string will be split into two when expanded.
-  set(ZENOHC_CARGO_FLAGS "--features=shared-memory zenoh/transport_serial")
+  if(ANDROID)
+    # shared-memory (the shared_memory crate) hard-fails to compile on Android
+    # (compile_error!("shared_memory isnt implemented for this platform...")).
+    # zenoh/transport_serial is untested there too. Stick to zenoh-c's default
+    # cargo features when cross-compiling.
+    set(ZENOHC_CARGO_FLAGS "")
+  else()
+    set(ZENOHC_CARGO_FLAGS "--features=shared-memory zenoh/transport_serial")
+  endif()
 
   include(get_cargo_version.cmake)
   get_cargo_version()
